### DISCUSSION TITLE
Multitenant / multisite questions

### DISCUSSION THREAD

**Discussion by undersound** [area17/twill#2171] [Q&A]
Couple of questions regarding multi-tenancy:

1. I succesfully enabled https://twill.io/docs/3.x/getting-started/configuration.html#multiple-subdomains-cms-routing. Now I want to make content domain specific so e.g. UserA logs in at admin.site1.test and only sees content which belongs to that domain. Any example or guidance is greatly appreciated.
2. Is it possible to use different databases per tenant/domain? Or do I need to use another package (looking at https://github.com/spatie/laravel-multitenancy for example)

Any examples or guidance is greatly appreciated

**Comment by ifox** [comment]
Hi @undersound! Quick answers for now:

1. You can enable Twill 3 permissions management for this. It will allow you to create groups of users with access to specified subdomains.

2. We have not explored this yet but I don't think another package would be required as we could simply dynamically change the database connection in the middleware that deals with the subdomain routing and configuration.

**Comment by undersound** [comment]
Thanks @ifox 

1. Could you eloborate a bit on how content will be glued to a specific domain/site? Let's say I have a module called 'pages' and userA creates "page 1" on admin.site1.test and userB creates "page 2" on admin.site2.test from this module 'pages'. How does content on e.g. admin.site1.test knows it only needs to show up there.

2. Thanks

**Comment by kallefrombosnia** [comment]
Not really my initial idea, but I will explain statement 1.

This is simple since you plan to use a multi-tenant setup with different databases per subdomain. Per that, content saved on website1 will only be shown on website1 since it will load DB connection1 (statement 2) via middleware.
But as @ifox said, this is not currently supported, for the time being, you might write your own middleware for that.

Note: when making migrations, you will also need to migrate on different connections, since Laravel migrates only on default connection when using `php artisan:migrate`. So you will need something like this `php artisan migrate:install --database=NAME_OF_CONNECTION`

What about auth, do managers/admins share the auth between subdomains or they will have separate authentication?

**Comment by undersound** [comment]
Thanks @kallefrombosnia  

In a multi database setup the content is ofcourse automatically seperated. I was curious though how content would be seperated in a single database setup by leveraging the twill 3 permissions management system if at all possible.

I just did a test with https://github.com/spatie/laravel-multitenancy in combination with Twill and it is working quite well at first glance. Database switching is determined by Spatie\Multitenancy\TenantFinder\DomainTenantFinder. There is also an artisan command which runs on all tenants. E.g. `php artisan tenants:artisan "migrate"` 

This reads all rows from the tenants table in the landlord database.
